### PR TITLE
Allow multiphase modules to be re-imported

### PR DESCRIPTION
**Pull request by b-pass**
## Description

This fixes the exception reported in #5780 and gets things *close* to the behavior of single-phase module initialization.  When import is done on a single-phase module after a `del sys.modules[...]`, the single-phase module is completely reloaded and reinitialized.  

I could not get CPython will not do this to multi-phase modules (I tried a few different things).. Instead, this PR keeps a cache of fully created and initialized modules (in the interpreter state dict).  When creating a module, we look in the cache first and use the already created and initialized one if it is in there.  Otherwise we create a new one.  And then also before running the user's exec function we check in the cache.

So, with this PR, the difference is that single-phase modules are reinitialized after a `del`, and multi-phase modules silently return a cached handle to the previously initialized instance.  I think this difference is probably OK because doing a `del sys.modules[...]` is probably not very common, and even in this case it was done as part of a test setup.

## Suggested changelog entry:

* Fixed issue that caused ``PYBIND11_MODULE`` code to run again if the module was re-imported after being deleted from ``sys.modules``.

### CODE REVIEW

File: include/pybind11/detail/common.h [hunk 1]
```diff
@@ -455,7 +455,11 @@ PyModuleDef_Init should be treated like any other PyObject (so not shared across
     int PYBIND11_CONCAT(pybind11_exec_, name)(PyObject * pm) {                                    \
         try {                                                                                     \
             auto m = pybind11::reinterpret_borrow<::pybind11::module_>(pm);                       \
-            PYBIND11_CONCAT(pybind11_init_, name)(m);                                             \
+            if (!pybind11::detail::get_cached_module(pybind11::getattr(                           \
+                    pybind11::getattr(m, "__spec__", m), "name", pybind11::none()))) {            \
```
Comment (rwgk) [inline include/pybind11/detail/common.h]: I'm worried about the `getattr(m, "__spec__", m)` fallback to `m` here masking bugs (and then leading to confusion).

If I can believe ChatGPT, for Python 3.8+, pybind11 extension modules will always have the `__spec__` attribute.

WDYT about "stricter is safer" here? Concretely, `pybind11_fail()` if the `__spec__` attribute doesn't exist.

File: include/pybind11/pybind11.h [hunk 2]
```diff
@@ -1329,9 +1329,56 @@ inline void *multi_interp_slot(F &&, O &&...o) {
...
+    }
+    return cache[nameobj].ptr();
+}
+
+/*
+Add successfully initialized a module object to the internal cache.
+*/
+static inline void cache_completed_module(pybind11::object mod) {
+    auto nameobj = getattr(getattr(mod, "__spec__", mod), "name", none());
+    dict state = detail::get_python_state_dict();
```
Comment (rwgk) [inline include/pybind11/pybind11.h:1354]: `nameobj` could be `None`: `pybind11_fail` in that case?

File: include/pybind11/pybind11.h [hunk 3]
```diff
@@ -1329,9 +1329,56 @@ inline void *multi_interp_slot(F &&, O &&...o) {
...
+        state["__pybind11_module_cache"] = dict();
+    }
+    state["__pybind11_module_cache"][nameobj] = mod;
+}
+
+/*
+A Py_mod_create slot function which will return the previously created module from the cache if one
+exists, and otherwise will create a new module object.
+*/
+static inline PyObject *cached_create_module(PyObject *spec, PyModuleDef *) {
```
Comment (rwgk) [inline include/pybind11/pybind11.h]: I believe this needs to be `extern "C"`, because the function pointer needs to have `C` linkage.

File: include/pybind11/pybind11.h [hunk 4]
```diff
@@ -1329,9 +1329,56 @@ inline void *multi_interp_slot(F &&, O &&...o) {
...
+    (void) &cache_completed_module; // silence unused-function warnings, it is used in a macro
+
+    auto nameobj = reinterpret_steal<str>(PyObject_GetAttrString(spec, "name"));
```
Comment (rwgk) [inline include/pybind11/pybind11.h]: I think we need to check `PyErr_Occurred()` here?

File: tests/test_modules.py [hunk 5]
```diff
@@ -71,6 +71,20 @@ def test_importing():
...
 
 
+def test_reimport():
+    import sys
+
+    import pybind11_tests as x
+
+    del sys.modules["pybind11_tests"]
+
+    # if something is wrong, this will throw import error ... otherwise nothing happens.
```
Comment (rwgk) [inline tests/test_modules.py]: We cannot really promise that it'll be an import error, and retrieving the module object from the cache (IIUC?) isn't exactly "nothing"...

How about:

```
# exercise pybind11::detail::get_cached_module()
```

File: tests/test_modules.py [hunk 6]
```diff
@@ -71,6 +71,20 @@ def test_importing():
...
+    import pybind11_tests as y
+
+    assert x is y
+
+
+@pytest.mark.xfail("env.GRAALPY", reason="TODO should be fixed on GraalPy side")
```
Comment (rwgk) [inline tests/test_modules.py]: @msimacek for visibility

**Review by rwgk — COMMENTED** [review]
Here is my related ChatGPT conversation:

https://chatgpt.com/share/688c74a5-2e08-8008-9c30-daf546a2f9a8

I'm still pretty uncertain about the `static inline` question, even though ChatGPT strongly suggests keeping the `static`. We have many helper functions that are not `static`.

File: include/pybind11/detail/common.h (hunk 1 above)
Comment (b-pass) [reply include/pybind11/detail/common.h]: Done

File: include/pybind11/pybind11.h (hunk 3 above)
Comment (b-pass) [reply include/pybind11/pybind11.h]: I don't think it does, CPython doesn't care what the name of the symbol is only what the pointer is as supplied by the slot.

File: include/pybind11/pybind11.h [hunk 7]
```diff
@@ -1329,18 +1329,72 @@ inline void *multi_interp_slot(F &&, O &&...o) {
...
+        return nullptr;
+    }
+
+    auto *cached = get_cached_module(nameobj);
+    if (cached) {
+        Py_INCREF(cached);
+        return cached;
+    }
+
+    Py_INCREF(nameobj.ptr());
```
Comment (rwgk) [inline include/pybind11/pybind11.h]: I think this `Py_INCREF()` isn't needed.

I looked here to come to that conclusion:

* https://docs.python.org/3/c-api/module.html#c.PyModule_NewObject
* https://github.com/python/cpython/blob/fe0e921817a7f96c62c91085884ab910859328ce/Objects/moduleobject.c#L126-L140
* https://github.com/python/cpython/blob/fe0e921817a7f96c62c91085884ab910859328ce/Objects/moduleobject.c#L64-L87
* https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItem

WDYT?

File: include/pybind11/detail/common.h [hunk 8]
```diff
@@ -455,7 +455,12 @@ PyModuleDef_Init should be treated like any other PyObject (so not shared across
     int PYBIND11_CONCAT(pybind11_exec_, name)(PyObject * pm) {                                    \
         try {                                                                                     \
             auto m = pybind11::reinterpret_borrow<::pybind11::module_>(pm);                       \
-            PYBIND11_CONCAT(pybind11_init_, name)(m);                                             \
+            auto mod_spec_name = pybind11::getattr(                                               \
+                pybind11::getattr(m, "__spec__", pybind11::none()), "name", pybind11::none());    \
```
Comment (rwgk) [inline include/pybind11/detail/common.h]: I saw your commit message for 23de0c315d391a47c6dd5b9fca56f0a8ff47d6ab:

> if the module doesn't have the right properties to check the cache then just allow exec to run.

That seem like a good defensive approach. We don't want to make too many assumptions about when Python does what. — Does that match what you had in mind?

However, when we get to `cache_completed_module()` three lines down from here, I'd expect that the module object must be complete. To not have silent failures, I'd want to try getting `m.__spec__.name` again, if it wasn't available here, and `pybind11_fail` if it still isn't available.

However again, do we really need to be defensive here? To be defensive and safe (no silent failures) adds complexity. Although not much.

If it was just me: I'd simply `pybind11_fail` here if `mod_spec_name` is `None`, and wait to see if we actually get any pushback.

Unless I'm missing something?

File: tests/test_modules.py (hunk 6 above)
Comment (rwgk) [reply tests/test_modules.py]: In case this TODO survives for some time, it'll be useful to have a pointer back here:

```
reason="TODO should be fixed on GraalPy side (see PR #5782)"
```

File: include/pybind11/detail/common.h (hunk 8 above)
Comment (b-pass) [reply include/pybind11/detail/common.h]: Ok.  Exceptions here get caught and turned into ImportError, so we can just use the throwing versions of getattr.
